### PR TITLE
DOC: explicitly define numpy.datetime64 semantics

### LINKED ISSUE #20675: DOC: "datetime64" explicilty define the calendar semantics as "proleptic Gregorian" with fixed length of day.
### Issue with current documentation:

The current implementation of `datetime64` is based on a [proleptic Gregorian calendar](https://en.wikipedia.org/wiki/Proleptic_Gregorian_calendar):
```
>>> np.datetime64('1582-10-15') - np.timedelta64(1, 'D')
numpy.datetime64('1582-10-14')
```
with [astronomical year numbering](https://en.wikipedia.org/wiki/Astronomical_year_numbering)
```
>>> np.datetime64('0001-01-01') - np.timedelta64(1, 'D')
numpy.datetime64('0000-12-31')
>>> np.datetime64('0000-01-01') - np.timedelta64(1, 'D')
numpy.datetime64('-001-12-31')
```
Both these important assumptions are left implicit and not stated explicitly.

### Idea or request for content:

I propose to clearly document right at the beginning of `doc/source/reference/arrays.datetime.rst` that `datetime64` computations are based on a proleptic Gregorian calendar with astronomical year numbering.

Moreover the fact that a day has a constant length of 86400 SI seconds is only cursorily mentioned when discussing Posix semantics and leap seconds. The documentation should stress this assumptions and its implications beyond the mere missing of leap seconds.

Of course I volunteer to open a DOC only PR.

### PR DESCRIPTION
**Pull request by miccoli**
Explicity state datetime64 semantics:
- adoption of proleptic Gregorian Calendar
- Astronomical year numbering
- 86400 s constant length of day.

Add some exmples of shortcomings of the above semnatics with ref. to
UTC timescale (missing leap seconds) and UT timescale (variable
legth of day).

Closes GH-20675

### CODE REVIEW

File: doc/source/reference/arrays.datetime.rst [hunk 1]
```diff
@@ -384,3 +411,69 @@ Some examples::
...
+     <https://en.wikipedia.org/wiki/Universal_Time>`_:
+
+      >>> a = np.datetime64("0000-01-01", "us")
+      >>> b = np.datetime64("1600-01-01", "us")
+      >>> b - a
+      numpy.timedelta64(50491123200000000,'us')
+
+     The computed results, `50491123200` seconds, is obtained as the elapsed
+     number of days (`584388`) times `86400` seconds; this is the number of
+     seconds of a clock in sync with earth rotation. The exact value in SI
```
Comment (charris) [inline doc/source/reference/arrays.datetime.rst:474]: I'm confused here, wouldn't a clock in sync with the earths rotation be UTC?

Comment (miccoli) [reply doc/source/reference/arrays.datetime.rst:474]: No, by defnition a clock in sync with earth rotation would be UT1 or, disregarding polar motion, simply UT. It is known that the earth is slowing down its rotation rate, therefore one UT1 second is not a uniform time scale. UTC on the contrary ticks at exactly one SI second rate, and therefore is "uniform". It is the stepping provided by the leap seconds that keeps UTC within 0.9s from UT1 and hence earth rotation.

UTC is a "realized" time scale, so it is not define before its inception (1972 with current rules, 1963 with a different stepping mechanism).

**Comment by miccoli** [comment]
This PR is slowly drifting into oblivion... probably because it is considered low priority.

Let me try an apology and clarify the motivation. First of all, numpy docs should explicitly state that the proleptic Gregorian calendar with astronomical year numbering is adopted.

1. Dates before 1582 are ambiguous. Quoting ISO 8601 (Date and time — Representations for information interchange)

    > Usage of the Gregorian calendar for identifying dates preceding its introduction (15 October 1582) should only be by mutual agreement of the communicating partners.

    This requires that the calendar system should be explicitly defined, as e.g. the Python [`datetime` docs](https://docs.python.org/3/library/datetime.html#date-objects) do.
    
1. Astronomical year numbering is not ambiguous (year zero and negative years do not exist in the BC/AD system used in historical year numbering) but casual users may wrongly assume that year 100BC is year -100.

Finally the assumption of 86400 second per year is at odds with current civil time usage, because UTC has positive and negative leap seconds. A brief remainder of this fact could be useful and is introduced by this PR.

The final part of the PR simply oultines the fact that  timestamping distant past event with µs granularity makes little sense, if the adopted timescale is not carefully discussed. In other words we have only an approximate knowledge of the UT timescale in the past (and also in the future, since earth rotation is impredictable with this accuracy) so that the achievable precision (given by the `datetime64` granularity) is several order of magnitude smaller than any sensible accuracy.

File: doc/source/reference/arrays.datetime.rst (hunk 1 above)
Comment (charris) [reply doc/source/reference/arrays.datetime.rst:474]: Maybe note that the value of second will change as the earth's rotation rate changes..

**Comment by charris** [comment]
Sorry for letting this slip by, I did intend to merge it with some clarifications.

**Comment by miccoli** [comment]
> Sorry for letting this slip by, I did intend to merge it with some clarifications.

No problem for the delay, this is just a minor correction to the docs. 

As what regards your request of clarification, I think that we can safely omit the last example, that shows the difference between an uniform SI second time scale (like TAI or [TT](https://en.wikipedia.org/wiki/Terrestrial_Time)) and an earth rotation timescale like UT. The discussion of TT vs UT1 is in fact beyond the mere `datetime64` semantics, and can be misleading for users not interested in astronomy. Moreover users interested in astronomy would use more specialized libraries, like astropy.

Let me know if you think that I should delete the last example.

**Comment by charris** [comment]
I think the difference between uniform seconds and earth rotation seconds is interesting and worth knowing about, it just needs to be clear why fixing a (solar?) day to 86400 seconds results in non-uniform seconds.

**Comment by charris** [comment]
Thanks @miccoli .

**Pull request merged**